inc/admin/options.php

```diff
@@ -88,6 +88,21 @@ function rocket_after_save_options( $oldvalue, $value ) {
 		rocket_clean_cache_busting();
 	}
 
+	if ( ! empty( $_POST ) &&
```

Comment:
Some issues here I think.
- Case 1:
```php
$oldvalue = [ 'cloudflare_email' => 'foo', 'cloudflare_key' => 'bar', 'cloudflare_id' => 'baz' ];
$value    = [ 'cloudflare_email' => '', 'cloudflare_key' => '', 'cloudflare_id' => '' ];
```
If any entry of `$value` is set but empty while those in `$oldvalue` are not, you get in the `if` statement. Then `rocket_is_api_keys_valid_cloudflare()` is fed with empty values. I don't know what will happen but it probably is not a good thing.
I imagine this will happen each time the user empty the fields, it will display an error message from Cloudflare.

- Case 2:
```php
$oldvalue = [ 'cloudflare_email' => 'foo', 'cloudflare_key' => 'bar', 'cloudflare_id' => 'baz' ];
$value    = [ 'cloudflare_email' => 'bar' ];
```
If any entry of `$value` differs from `$oldvalue` while any other is not set, you also get in. Then php notices are triggered.

- This might be useful at some point:
```php
$has_values = ! empty( $value['cloudflare_email'] ) && ! empty( $value['cloudflare_api_key'] ) && ! empty( $value['cloudflare_zone_id'] );

if ( ! empty( $_POST ) && $has_values && (
	empty( $oldvalue['cloudflare_email'] ) || ( isset( $oldvalue['cloudflare_email'] ) && $oldvalue['cloudflare_email'] !== $value['cloudflare_email'] ) ||
	...
```

Reply:
Empty values are handled in `rocket_is_api_keys_valid_cloudflare()` function and will display proper error messages. It will not end up to CF API until all values are filled.

https://github.com/wp-media/wp-rocket/blob/cc93d257fb625a439f76c1e1b31dadd7a201f912/inc/functions/cloudflare.php#L21-L46

Reply:
@Screenfeed For case 2, the values are sanitized in the `sanitize_callback` function before going through this function, making sure they are always set even if empty in the form:

https://github.com/wp-media/wp-rocket/blob/1e915944d0bf9226e49a3f6944a7f88b33c8fc1e/inc/classes/admin/settings/class-settings.php#L359

Reply:
OK, thanks for your answers.
So I guess this means `isset( $oldvalue['cloudflare_email'], $value['cloudflare_email'] )` can be reduced to `isset( $oldvalue['cloudflare_email'] )`, right?
But then, the case where it is not set is not handled, which would be:
```php
! isset( $oldvalue['cloudflare_email'] ) && $value['cloudflare_email'] ||
isset( $oldvalue['cloudflare_email'] ) && $oldvalue['cloudflare_email'] !== $value['cloudflare_email'] ||
```
And for better readability:
```php
$oldvalue['cloudflare_email'] = ! empty( $oldvalue['cloudflare_email'] ) ? $oldvalue['cloudflare_email'] : '';
...
if ( ...
	$oldvalue['cloudflare_email'] !== $value['cloudflare_email'] ||
	...
```

Reply:
In theory it's never not set either, but you know how it is 😅 

I like the tertiary operator before the condition for the 3 `oldvalue`, that way we should be safe